compiler/rustc_data_structures/src/sync.rs

```diff
@@ -18,13 +18,30 @@
 //! depending on the value of cfg!(parallel_compiler).
 
 use crate::owning_ref::{Erased, OwningRef};
+use std::cell::{RefCell, RefMut, SyncUnsafeCell};
 use std::collections::HashMap;
 use std::hash::{BuildHasher, Hash};
 use std::ops::{Deref, DerefMut};
 use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
 
 pub use std::sync::atomic::Ordering;
 pub use std::sync::atomic::Ordering::SeqCst;
+use std::sync::{Mutex, MutexGuard};
+
+static PARALLEL: SyncUnsafeCell<bool> = SyncUnsafeCell::new(false);
+
+#[inline]
+pub fn is_parallel() -> bool {
+    // SAFETY: `PARALLEL` only modified by the unsafe function `set_parallel`, so
+    // this function is safe if `set_parallel` be used correctly.
+    unsafe { *PARALLEL.get() }
+}
+
+// Only set by the `-Z threads` compile option
+pub unsafe fn set_parallel() {
+    let p = SyncUnsafeCell::raw_get(&PARALLEL as *const _);
+    *p = true;
+}
```

Comment:
First of all, it would be great to have a doc comment here, especially given that this is an `unsafe` function. Second of all, at first glance it seems like this can be more simply written as `*PARALLEL.get() = true`, am I missing something? Lastly, is `is_parallel` hot? Can we use an `AtomicUsize` instead?

Reply:
Even if it's a little hot, it's unlikely that an atomic integer will have a performance impact, since this is just reading from it.

Reply:
Thanks for the review! `PARALLEL` will only be set once, so I want to take advantage of this to minimize the cost of reading it. `with_context_opt` might be hot, but I doubt the necessary to check thread safety here. Except this `is_parallel()` is not hot, since it is only used in relatively top-level logic to determine whether to switch to a parallel environment.

Reply:
You should first benchmark it before going for the more unsafe variant. Atomics have no to minimal overhead depending on the exact use and ordering (which I think can be relaxed here because we don't need to sync any other writes?).

Reply:
OK. I changed to AtomicBool instead. Can you help run a perf?Thanks!